### PR TITLE
Move browser opening logic in `Builder`

### PR DESCRIPTION
**Pull request by emilyalbini**
This allows `open()` to be called from other places in bootstrap (I need this for Ferrocene, as we keep our custom steps in `src/bootstrap/ferrocene`), and it simplifies the callers by moving the `was_invoked_explicitly` check into the function.

### CODE REVIEW

File: src/bootstrap/doc.rs [hunk 1]
```diff
@@ -384,10 +370,7 @@ impl Step for Standalone {
 
         // We open doc/index.html as the default if invoked as `x.py doc --open`
         // with no particular explicit doc requested (e.g. library/core).
-        if builder.paths.is_empty() || builder.was_invoked_explicitly::<Self>(Kind::Doc) {
-            let index = out.join("index.html");
-            open(builder, &index);
-        }
+        builder.maybe_open_in_browser::<Self>(out.join("index.html"));
```
Comment (jyn514) [inline src/bootstrap/doc.rs:373]: Hmm, I think the `paths.is_empty()` check is now missing, right? Can you test whether `x.py doc --open` with no arguments still opens the index page?

Instead I think you could keep the `invoked_explicitly` check for this function only and then call `open_in_browser` directly.

Comment (pietroalbini) [reply src/bootstrap/doc.rs]: Good catch, missed it.

**Comment by jyn514** [comment]
bors is asleep again :weary:

@bors r+ rollup

**Pull request merged**